### PR TITLE
Add phpstan-phpunit phpstan-strict-rules to phpstan analysis

### PR DESCRIPTION
**Pull request by phil-davis**
see my comments - fixes a few edge-cases when a URI component might be the string "0" (path, username, query, fragment...). These are not really common in everyday use!

Gets the strict phpstan rules passing.

### CODE REVIEW

**Comment by phil-davis** [comment]
```
------ ---------------------------------------------------------------------- 
  Line   lib/functions.php                                                     
 ------ ---------------------------------------------------------------------- 
  65     Short ternary operator is not allowed. Use null coalesce operator if  
         applicable or consider using long ternary.                            
  90     Construct empty() is not allowed. Use more strict comparison.         
  90     Construct empty() is not allowed. Use more strict comparison.         
  90     Construct empty() is not allowed. Use more strict comparison.         
  116    Construct empty() is not allowed. Use more strict comparison.         
  144    Construct empty() is not allowed. Use more strict comparison.         
  152    Construct empty() is not allowed. Use more strict comparison.         
  254    Construct empty() is not allowed. Use more strict comparison.         
  256    Construct empty() is not allowed. Use more strict comparison.         
  259    Construct empty() is not allowed. Use more strict comparison.         
  264    Construct empty() is not allowed. Use more strict comparison.         
  268    Construct empty() is not allowed. Use more strict comparison.         
  273    Construct empty() is not allowed. Use more strict comparison.         
  276    Construct empty() is not allowed. Use more strict comparison.         
  279    Construct empty() is not allowed. Use more strict comparison.         
 ------ ---------------------------------------------------------------------- 
```

I will have a go at refactoring these. But it often seems a bit painful to remove use of `empty()` - need to check if something is `null`, the empty string etc.

**Comment by staabm** [comment]
if you don't see value in changing existing code, or don't have time todo it now, feel free to [create a baseline for them](https://phpstan.org/user-guide/baseline).

File: lib/functions.php [hunk 1]
```diff
@@ -62,7 +62,10 @@ function resolve(string $basePath, string $newPath): string
             $path .= '/'.$delta['path'];
         }
     } else {
-        $path = $base['path'] ?: '/';
+        $path = $base['path'] ?? '/';
+        if ('' === $path) {
+            $path = '/';
+        }
```
Comment (phil-davis) [inline lib/functions.php:67]: phpstan strict rules do not like use of the "?:" operator. It suggests using the "??" null coalesce operator or the full ternary operator. Actually that is "a good thing", because in this case if `$base['path']` is the string "0" then `?:` sees that as false and takes the "/", which results in the "0" path disappearing from the URI - that is an existing (very edge case) bug.

I added unit test cases to cover various permutations of the base path being "0" "a" or an empty string. One of the "0" cases fails prior to this code change, and now passes.

The interesting different cases for `$base['path']` are when it is not set (`null`) - that is covered by the `??` null coalesce operator - "0" which now works, and the empty string - which now has a specific check.

The wonders of PHP!

File: lib/functions.php [hunk 2]
```diff
@@ -251,32 +254,32 @@ function build(array $parts): string
     $uri = '';
 
     $authority = '';
-    if (!empty($parts['host'])) {
+    if (isset($parts['host'])) {
         $authority = $parts['host'];
-        if (!empty($parts['user'])) {
+        if (isset($parts['user'])) {
```
Comment (phil-davis) [inline lib/functions.php:262]: These changes from `!empty()` to `isset()` now allow the value to be "0".
I added unit test cases for ways that I could see that the value of these things might be "reasonably" passed in as the value "0". For example, a user "name" might be the string "0"! Or the query string could be just `?0` (very edge case - an application that just sends a number in the query string). Or the fragment could be `#0`. All of those unit test cases fail with the `!empty()` check, and pass with `isset()`

**Comment by phil-davis** [comment]
@staabm please review.

File: lib/functions.php [hunk 3]
```diff
@@ -85,13 +88,17 @@ function resolve(string $basePath, string $newPath): string
 
     // If the source url ended with a /, we want to preserve that.
     $newParts['path'] = 0 === strpos($path, '/') ? $path : '/'.$path;
-    if ($delta['query']) {
+    // From PHP 8, no "?" query at all causes 'query' to be null.
+    // An empty query "http://example.com/foo?" causes 'query' to be the empty string
+    if (null !== $delta['query'] && '' !== $delta['query']) {
         $newParts['query'] = $delta['query'];
-    } elseif (!empty($base['query']) && empty($delta['host']) && empty($delta['path'])) {
+    } elseif (isset($base['query']) && !isset($delta['host']) && !isset($delta['path'])) {
```
Comment (staabm) [inline lib/functions.php]: per return type

https://github.com/phil-davis/uri/blob/3a0ca956c569d9c4a4782de785b696df89378089/lib/functions.php#L187

`host` and `path` are always set. is this `!isset` effectively a `!== null` check?

File: lib/functions.php [hunk 4]
```diff
@@ -113,7 +120,7 @@ function normalize(string $uri): string
 {
     $parts = parse($uri);
 
-    if (!empty($parts['path'])) {
+    if (isset($parts['path'])) {
```
Comment (staabm) [inline lib/functions.php]: same as above?

```suggestion
    if ($parts['path'] !== null) {
```

and a few times more below

**Review by staabm — APPROVED** [review]
lgtm - except a few nits

thank you

File: lib/functions.php (hunk 3 above)
Comment (phil-davis) [reply lib/functions.php]: yes, I suppose it amounts to that. `$base['somekey']` or `$delta['somekey']` might not "have anything" - the keys may not be set (they might "return null" when accessed)

File: lib/functions.php (hunk 4 above)
Comment (phil-davis) [reply lib/functions.php]: yes, I should be consistent - `isset` and `!== null` are essentially the same in these places.
I will adjust...

File: lib/functions.php (hunk 3 above)
Comment (phil-davis) [reply lib/functions.php]: Changing to `null !==` - coming in next commit.

File: lib/functions.php (hunk 4 above)
Comment (phil-davis) [reply lib/functions.php]: Change coming...

**Comment by phil-davis** [comment]
I broke a heap of unit tests - I will look...

File: lib/functions.php [hunk 5]
```diff
@@ -378,7 +381,14 @@ function ($matches) {
     if ('///' === substr($uri, 0, 3)) {
         // The triple slash uris are a bit unusual, but we have special handling
         // for them.
-        $result['path'] = substr($uri, 2);
+        /**
+         * We know that this is a string, because the first 3 characters are '///'.
+         * So we can always do a sub-string starting from offset 2 (the 3rd '/').
+         *
+         * @var string $x
```
Comment (staabm) [inline lib/functions.php]: [inline `@var` should be prevented when possible.](https://phpstan.org/blog/phpstan-1-10-comes-with-lie-detector#validate-inline-phpdoc-%40var-tag-type)

I wonder why its needed in the first place:
https://phpstan.org/r/e1d7178d-1eff-4a4e-bb0c-cd2d9768c767
-> looks good to me.

Comment (phil-davis) [reply lib/functions.php]: phpstan kept thinking that it could be `false`. but we know better - we can always do substr from offset 2 at this point.

It seems to understand on PHP 8 but does not understand on PHP 7.4.

Maybe we should run phpstan with PHP 8.? in CI, and not with PHP 7.4. But we still want our code to be compliant with PHP 7.4

Comment (staabm) [reply lib/functions.php]: in 7.4 it can return false. 

please just add a `if ($x === false) throw new \RuntimeException('The string cannot be false');` 
instead of the inline-`@var`

that way we see immediately whether our assumption is correct

**Review by staabm — APPROVED** [review]
lgtm

File: lib/functions.php (hunk 5 above)
Comment (phil-davis) [reply lib/functions.php]: done

File: lib/functions.php (hunk 2 above)
Comment (phil-davis) [reply lib/functions.php:262]: Note: we need to use `isset()` because the `$parts` array passed to `build()` may not have all the component keys set.

In other places we could compare directly to `null` because `resolve()`, `normalize()` etc. we guarantee that all the URI components are actually set in the array (and "missing" components are set to value `null`
